fw/ringbuffer.c

```diff
@@ -0,0 +1,248 @@
+/**
+ *		Tempesta FW
+ *
+ * Handling ring buffers mapped to userspace.
+ *
+ * Copyright (C) 2024 Tempesta Technologies, Inc.
+ *
+ * This program is free software; you can redistribute it and/or modify it
+ * under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2 of the License,
+ * or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or
+ * FITNESS FOR A PARTICULAR PURPOSE.
+ * See the GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License along with
+ * this program; if not, write to the Free Software Foundation, Inc., 59
+ * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
+ */
+
+#include "ringbuffer.h"
+#include "lib/str.h"
+#include <linux/types.h>
+#include <asm/io.h>
+#include <linux/log2.h>
+#include <linux/cpumask.h>
+
+typedef struct {
+	TfwRingbuffer *rb;
+	struct page *pg;
+	atomic_t unmapped;
+} TfwRingbufferArea;
+
+#define AREA_SIZE(size) (size + TFW_RINGBUFFER_DATA_OFFSET)
+#define RBA_AREA_SIZE(rba) AREA_SIZE((rba)->rb->size)
+
+static DEFINE_PER_CPU(TfwRingbufferArea, rb_area);
+static bool proc_file_is_open;
+
+int
+tfw_ringbuffer_write(TfwStr **strs, unsigned int count)
```

Comment:
I am not sure but it seems that we can use just
```
int
tfw_ringbuffer_write(TfwStr *str)
```
because TfwStr can contains a lot of TfwStr as chunks.

Reply:
We agreed on the call that this API introduces unwanted overhead on forming an array, so probably we should just provide a raw memory and a commit call, which moves the `tail`